geode-core/src/upgradeTest/java/org/apache/geode/internal/cache/rollingupgrade/RollingUpgradeWithGfshDUnitTest.java

```diff
@@ -0,0 +1,160 @@
+    result.removeIf(s -> TestVersion.compare(s, "1.10.0") < 0);
+    return result;
+  }
+
+  @Rule
+  public GfshRule oldGfsh;
+
+  @Rule
+  public GfshRule currentGfsh = new GfshRule();
+
+  @Rule
+  public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  public RollingUpgradeWithGfshDUnitTest(String version) {
+    oldVersion = version;
+    oldGfsh = new GfshRule(oldVersion);
+  }
+
+  @Test
+  public void testRollingUpgradeWithDeployment() throws Exception {
+    int locatorPort = portSupplier.getAvailablePort();
+    int locatorJmxPort = portSupplier.getAvailablePort();
+    int locator2Port = portSupplier.getAvailablePort();
+    int locator2JmxPort = portSupplier.getAvailablePort();
+    int server1Port = portSupplier.getAvailablePort();
+    int server2Port = portSupplier.getAvailablePort();
+
+    GfshExecution startupExecution =
+        GfshScript.of(startLocatorCommand("loc1", locatorPort, locatorJmxPort, -1))
+            .and(startLocatorCommand("loc2", locator2Port, locator2JmxPort, locatorPort))
+            .and(startServerCommand("server1", server1Port, locatorPort))
+            .and(startServerCommand("server2", server2Port, locatorPort))
+            .and(deployDirCommand())
+            .execute(oldGfsh);
+
+    // doing rolling upgrades
+    oldGfsh.stopLocator(startupExecution, "loc1");
+    GfshScript.of(startLocatorCommand("loc1", locatorPort, locatorJmxPort, locator2Port))
+        .execute(currentGfsh);
+    verifyListDeployed(locatorPort);
+
+    oldGfsh.stopLocator(startupExecution, "loc2");
+    GfshScript.of(startLocatorCommand("loc2", locator2Port, locator2JmxPort, locatorPort))
+        .execute(currentGfsh);
+    verifyListDeployed(locator2Port);
+
+    // make sure servers can do rolling upgrade too
+    oldGfsh.stopServer(startupExecution, "server1");
+    GfshScript.of(startServerCommand("server1", server1Port, locatorPort)).execute(currentGfsh);
+
+    oldGfsh.stopServer(startupExecution, "server2");
+    GfshScript.of(startServerCommand("server2", server2Port, locatorPort)).execute(currentGfsh);
+  }
+
+  private void verifyListDeployed(int locatorPort) {
+    GfshExecution list_deployed = GfshScript.of("connect --locator=localhost[" + locatorPort + "]")
+        .and("list deployed").execute(currentGfsh);
+    assertThat(list_deployed.getOutputText()).contains("DeployCommandsDUnit1.jar")
+        .contains("server1").contains("server2");
+    currentGfsh.execute("disconnect");
+  }
+
+  private String deployDirCommand() throws IOException {
+    ClassBuilder classBuilder = new ClassBuilder();
+    File jarsDir = tempFolder.newFolder();
+    String jarName1 = "DeployCommandsDUnit1.jar";
+    File jar1 = new File(jarsDir, jarName1);
+    String class1 = "DeployCommandsDUnitA";
+    classBuilder.writeJarFromName(class1, jar1);
+    return "deploy --dir=" + jarsDir.getAbsolutePath();
+  }
+
+  protected String startServerCommand(String name, int port, int connectedLocatorPort) {
```

Comment:
Is there a subclass? I'd like to see these `protected` methods be `private` if possible.